modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java

```diff
@@ -654,12 +655,22 @@ private boolean isSafeLossPolicy() {
                 if (prj != null || part != null)
                     return nodes(cctx, prj, part);
 
-                if (cctx.affinityNode())
+                if (cctx.affinityNode() && !cctx.topology().localPartitionMap().hasMovingPartitions())
                     return Collections.singletonList(cctx.localNode());
 
                 Collection<ClusterNode> affNodes = nodes(cctx, null, null);
 
-                return affNodes.isEmpty() ? affNodes : Collections.singletonList(F.rand(affNodes));
+                List<ClusterNode> nodes = new ArrayList<>(affNodes);
+
+                Collections.shuffle(nodes);
+
+                for (ClusterNode node : nodes) {
+                    if (!cctx.topology().partitions(node.id()).hasMovingPartitions()) {
```

Comment:
Why do you think it will cause duplicates?
I supposed if I return all affinity nodes it will try to use only primary partitions in this case, like if we worked with PARTITIONED cache, correct me if I'm wrong.

Reply:
I meant for each partition distributed query manager will choose only primary node for this partition, thus no duplicate results possible. Otherwise if I'll return one affinity in such case, it certainly will lead to incomplete results. Do you agree?

Reply:
I can see from `GridCacheQueryManager#scanIterator` backup partitions will be always scanned for REPLICATED caches.

Reply:
Also I dug a little bit deeper and found some more problems.
First, there is a race between getting nodes and inspecting partitions. Consequently `cctx.topology().partitions(node.id())` can return `null` here. I suppose using single topology version here can help.
Second, LOST partitions are not handled properly. It is possible that `hasMovingPartitions` returns `false` when there are LOST partitions. Unpleasant, but in my mind can be fixed in a separate ticket about ScanQuery and LOST partitions.

Good news here are that if there is no LOST partitions it seems there is an invariant for REPLICATED caches that on a specific topology version at least one node owns all partitions.

Reply:
Hi, Ivan. I agree LOST partitions case better to solve in another ticket.
About the race - I did not realize how to use topology version here, but it looks like I solved the race with `topology.readLock()`, how do you like it?